datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -17,6 +17,7 @@
 
 //! Coercion rules for matching argument types for binary operators
```

Comment:
I renamed `binary_rule.rs` to type_coercion/binary.rs` to better explain what it is for. I have always been confused about the name `binary_rule`

I think this module has a lot more code than for binary operators so it can be split / cleaned up more but I felt this is a good start